pkg/objstore/s3/s3.go

```diff
@@ -47,46 +48,64 @@ type Bucket struct {
 type Config struct {
 	Bucket       string
 	Endpoint     string
-	AccessKey    string
-	SecretKey    string
+	Credentials  *credentials.Credentials
 	Insecure     bool
 	SignatureV2  bool
 	SSEEnprytion bool
 }
 
 // RegisterS3Params registers the s3 flags and returns an initialized Config struct.
 func RegisterS3Params(cmd *kingpin.CmdClause) *Config {
-	var s3config Config
+
+	var (
+		s3Config  Config
+		accessKey string
+		secretKey string
+	)
 
 	cmd.Flag("s3.bucket", "S3-Compatible API bucket name for stored blocks.").
-		PlaceHolder("<bucket>").Envar("S3_BUCKET").StringVar(&s3config.Bucket)
+		PlaceHolder("<bucket>").Envar("S3_BUCKET").StringVar(&s3Config.Bucket)
 
 	cmd.Flag("s3.endpoint", "S3-Compatible API endpoint for stored blocks.").
-		PlaceHolder("<api-url>").Envar("S3_ENDPOINT").StringVar(&s3config.Endpoint)
+		PlaceHolder("<api-url>").Envar("S3_ENDPOINT").StringVar(&s3Config.Endpoint)
 
 	cmd.Flag("s3.access-key", "Access key for an S3-Compatible API.").
-		PlaceHolder("<key>").Envar("S3_ACCESS_KEY").StringVar(&s3config.AccessKey)
+		PlaceHolder("<access-key>").StringVar(&accessKey)
 
-	s3config.SecretKey = os.Getenv("S3_SECRET_KEY")
+	cmd.Flag("s3.secret-key", "Secrety key for an S3-Compatible API.").
+		PlaceHolder("<secret-key>").StringVar(&secretKey)
 
 	cmd.Flag("s3.insecure", "Whether to use an insecure connection with an S3-Compatible API.").
-		Default("false").Envar("S3_INSECURE").BoolVar(&s3config.Insecure)
+		Default("false").Envar("S3_INSECURE").BoolVar(&s3Config.Insecure)
 
 	cmd.Flag("s3.signature-version2", "Whether to use S3 Signature Version 2; otherwise Signature Version 4 will be used.").
-		Default("false").Envar("S3_SIGNATURE_VERSION2").BoolVar(&s3config.SignatureV2)
+		Default("false").Envar("S3_SIGNATURE_VERSION2").BoolVar(&s3Config.SignatureV2)
 
 	cmd.Flag("s3.encrypt-sse", "Whether to use Server Side Encryption").
-		Default("false").Envar("S3_SSE_ENCRYPTION").BoolVar(&s3config.SSEEnprytion)
-
-	return &s3config
+		Default("false").Envar("S3_SSE_ENCRYPTION").BoolVar(&s3Config.SSEEnprytion)
+
+	if accessKey == "" || secretKey == "" {
+		creds := credentials.NewChainCredentials(
+			[]credentials.Provider{
+				&credentials.EnvAWS{},
+				&credentials.IAM{},
+			})
+		s3Config.Credentials = creds
+	} else {
+		if s3Config.SignatureV2 {
+			s3Config.Credentials = credentials.NewStaticV2(accessKey, secretKey, "")
+		} else {
+			s3Config.Credentials = credentials.NewStaticV4(accessKey, secretKey, "")
+		}
+	}
+	return &s3Config
 }
 
 // Validate checks to see if mandatory s3 config options are set.
 func (conf *Config) Validate() error {
 	if conf.Bucket == "" ||
 		conf.Endpoint == "" ||
-		conf.AccessKey == "" ||
-		conf.SecretKey == "" {
+		conf.Credentials == nil {
 		return errors.New("insufficient s3 configuration information")
 	}
 	return nil
```

Comment:
Not sure if it belongs in this PR, but can we try to differentiate between no config and a missing/invalid parameter?
Now there is no way store/sidecar can't tell if Validate() fails because one parameter is missing, or s3 is not used at all and the error given to the user when forgetting one parameter is misleading.

It would require some changes in the store and sidecar commands as well, so maybe it doesn't belong in this PR.

Reply:
Speaking of error handling, in the for loop in the Iter() function, it would be nice to 
check if object.Err is non-nil and return it if is. Otherwise 401 responses may get
swallowed.

There may be similar issues in the other functions, but that's the one that bit me.

Reply:
True, other than 401 being authentication error :) But yea, should probably go in another PR